include/zephyr/kernel/obj_core.h

```diff
@@ -0,0 +1,353 @@
+ *
+ * Initializing the kernel object core associates it with the specified
+ * kernel object type.
+ *
+ * @param obj_core Pointer to the kernel object to initialize
+ * @param type Pointer to the kernel object type
+ */
+void k_obj_core_init(struct k_obj_core *obj_core, struct k_obj_type *type);
+
+/**
+ * @brief Link the kernel object to the kernel object type list
+ *
+ * A kernel object can be optionally linked into the kernel object type's
+ * list of objects. A kernel object must have been initialized before it
+ * can be linked. Linked kernel objects can be traversed and have information
+ * extracted from them by system tools.
+ *
+ * @param obj_core Pointer to the kernel object
+ */
+void k_obj_core_link(struct k_obj_core *obj_core);
+
+/**
+ * @brief Automatically link the kernel object after initializing it
+ *
+ * A useful wrapper to both initialize the core of the kernel object and
+ * automatically link it into the kernel object type's list of objects.
+ *
+ * @param obj_core Pointer to the kernel object to initialize
+ * @param type Pointer to the kernel object type
+ */
+void k_obj_core_init_and_link(struct k_obj_core *obj_core,
+			      struct k_obj_type *type);
+
+/**
+ * @brief Unlink the kernel object from the kernel object type list
+ *
+ * Kernel objects can be unlinked from their respective kernel object type
+ * lists. If on a list, it must be done at the end of the kernel object's life
+ * cycle.
+ *
+ * @param obj_core Pointer to the kernel object
+ */
+void k_obj_core_unlink(struct k_obj_core *obj_core);
+
+/** @} */
+
+/**
+ * @defgroup obj_core_stats_apis Object Core Statistics APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+#ifdef CONFIG_OBJ_CORE_STATS
+/**
+ * @brief Initialize the object type's stats descriptor
+ *
+ * This routine initializes the object type's stats descriptor.
+ *
+ * @param type Pointer to the object type
+ * @param stats_desc Pointer to the object core statistics descriptor
+ */
+static inline void k_obj_type_stats_init(struct k_obj_type *type,
+					 struct k_obj_core_stats_desc *stats_desc)
+{
+	type->stats_desc = stats_desc;
+}
+
+/**
+ * @brief Initialize the object core for statistics
+ *
+ * This routine initializes the object core to operate within the object core
+ * statistics framework.
+ *
+ * @param obj_core Pointer to the object core
+ * @param stats Pointer to the object's raw statistics
+ */
+static inline void k_obj_core_stats_init(struct k_obj_core *obj_core,
+					 void *stats)
+{
+	obj_core->stats = stats;
+}
+#endif
+
+/**
+ * @brief Register kernel object for gathering statistics
+ *
+ * Before a kernel object can gather statistics, it must be registered to do
+ * so. Registering will also automatically enable the kernel object to gather
+ * its statistics.
+ *
+ * @param obj_core Pointer to kernel object core
+ * @param stats Pointer to raw kernel statistics
+ * @param stats_len Size of raw kernel statistics buffer
+ *
+ * @return 0 on success, -errno on failure
```

Comment:
nit: there is an inconsistent use of `@return` / `@retval` throughout the file

Reply:
Fixing the inconsistency. I still have to learn more about the 'literalinclude' before I am comfortable with it.